# Исследование надежности заемщиков


## Шаг 1. Открываем таблицу и изучаем общую информацию о данных

**Импортируем библиотеку pandas. Считаем данные из csv-файла в датафрейм и сохраним в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**1.1. Выводим первые 20 строчек датафрейма `data` на экран.**

In [2]:
# используем метод head() для вывода первых 20-ти строк датафрейма
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**1.2. Выводим основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
# выводим основную информацию датафрейма с помощью метода info()
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Шаг 2. Предобработка данных

### 2.1. Удаление пропусков

**Выводим количество пропущенных значений для каждого столбца.**

In [4]:
# считаем сумму пропусков каждого столбца с помощью последовательного применения методов isna() и sum()
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Обнаружены пропущенные значения в столбцах с данными о трудовом стаже (days_employed - 2 174(10,10%)) и о доходе (total_income - 2 174(10,10%)).

Такая доля пропусков может существенно влиять на отдельные расчёты если пропуски в малых категориях заёмщиков (income_type)

Каждый из столбцов с пропущенными значениями отражает количественные переменные. Пропуски мы можем заполнить медианным значением каждой из групп (income_type). Это позволит нам сохранить важную информацию датафрейма.

Для этого сгруппируем потенциальных заёмщиков по типу занятости, рассчитаем медианное значение дохода каждой из групп и заменим им пропущенные значения.

**В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце обработаем на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
# Создаём функцию, которая заполнит все пропуски в столбце с доходами медианными занчениями групп по типу занятости

for t in data['income_type'].unique():
    
    '''
    на каждом шаге цикла с помощью атрибута loc выбираем строки,
    в которых в income_type текущий тип занятости (i) и есть пропуски в total_income 
    и записываем в них медианное значение total_income среди строк с текущим типом занятости (i)
    '''
        
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()
   
# проверяем, что все пропуски заполнены
data['total_income'].isna().sum()

0

### 2.2. Обработка аномальных значений

**В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
#К столбцу с данными о стаже применяем функцию модуля и избавляемся от аномалий
data['days_employed'] = data['days_employed'].abs()

**Для каждого типа занятости выводим медианное значение трудового стажа `days_employed` в днях.**

In [7]:
# Выводим медианные значения о трудовом стаже каждой из групп
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставляем их как есть.

**Выводим перечень уникальных значений столбца `children`.**

In [8]:
# применяем метод unique() к столбцу 'children'
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

В столбце 'children' обнаружено два аномальных значения: -1 и 20. Такое количество детей вряд ли может быть реальным. Их следует исключить во избежание неверных выводов.

**В столбце `children` есть два аномальных значения. Удали строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
# с помощью метода loc исключаем строки с аномальными значениями в столбце 'children' 
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Ещё раз выводим перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
# ещё раз применяем метод unique() к столбцу 'children' для проверки на отсутствие аномалий
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### 2.1. Удаление пропусков (продолжение)

**Заполним пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
# Создаём функцию, которая заполнит все пропуски в столбце со стажем медианными занчениями групп по типу занятости

for t in data['income_type'].unique():
    
    # на каждом шаге цикла с помощью атрибута loc выбираем строки,
    # в которых в days_employed текущий тип занятости (d) и есть пропуски в days_employed 
    # и записываем в них медианное значение total_income среди строк с текущим типом занятости (d)
    
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()
    
# проверяем, что все пропуски заполнены
data['days_employed'].isna().sum()

0

**Убедимся, что все пропуски заполнены. Проверим себя и ещё раз выведем количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
# проверяем, что пропуски заполнены с помощью последовательного применения методов isna() и sum()
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### 2.3. Изменение типов данных

**Заменим вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
# применяем метод astype(int) для указания целочисленного типа данных в столбце 'total_income'

data['total_income'] = data['total_income'].astype(int)

### 2.4. Обработка дубликатов

**Обработаем неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведём их к нижнему регистру.**

In [14]:
# заменим столбец 'education' исправленными значениями
data['education'] = data['education'].str.lower()

**Выведем на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалим их.**

In [15]:
# считаем дубликаты
data.duplicated().sum()

71

In [16]:
# удаляем дубликаты
data = data.drop_duplicates()

### 2.5. Категоризация данных

**На основании диапазонов, указанных ниже, создаём в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используем собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
# создаем функцию categorize_income() для категоризации заёмщиков в зависимости от уровня дохода
def categorize_income(income):
    
    """
    Присваивает категорию в зависимости от уровня дохода (total_income), используя правила:
    - 0–30000 — 'E';
    - 30001–50000 — 'D';
    - 50001–200000 — 'C';
    - 200001–1000000 — 'B';
    - 1000001 и выше — 'A'.
    """
    
    
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
# применяем функцию методом apply()
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Выводим на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
# с помощью метода unique() выведем все значения столбца 'purpose' - с указанием целей кредитования
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Создаём функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используем собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучим данные в столбце `purpose` и определим, какие подстроки помогут правильно определить категорию.**

In [20]:
# создаём функцию categorize_purpose() для категоризации заёмщиков в зависимости от целей кредитования

def categorize_purpose(row):
    
    """
    Присваивает категорию в зависимости от целей кредита, используя правила:
    
    - если строка содержит 'автом' присвоить категорию 'операции с автомобилем';
    - если строка содержит 'жил' или 'недвиж' присвоить категорию 'операции с недвижимостью';
    - если строка содержит 'свад' присвоить категорию 'проведение свадьбы';
    - если строка содержит 'образов' присвоить категорию 'получение образования'
    - во всех других случаях присвоить 'нет категории'
    """
    
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
# применяем функцию методом apply()
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуем данные и отвечаем на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:

'''
Для ответа на вопрос сгруппируем данные таблицы по количеству детей.

В таблице отразим данные по столбцу'debt' – "имел ли задолженность по возврату кредитов" с булевым типом данных. 
Добавим столбец-счётчик общего количества для расчёта доли должников в каждой из групп.

Чем выше значение ячейки, тем больше задолженностей по кредиту у анализируемой категории заёмщиков.
В столбце 'count' отразим данные счётчика.

Сделаем выводы о количестве заёмщиков с задолженностями каждой из групп по количеству детей.
'''

data_pivot = data.pivot_table(index=['children'], values='debt', aggfunc={'mean', 'sum'}) # Это старый вариант кода
data_pivot

mean   sum
children                
0         0.075438  1063
1         0.092346   444
2         0.094542   194
3         0.081818    27
4         0.097561     4
5         0.000000     0

In [23]:
# Создаю функцию 'display_pivot_table' для вывода сводной таблицы по любой из категорий датафрейма

def display_pivot_table(column_name):
    
    '''
    Функция формирует и выводит сводную таблицу по одному из сегментов (столбцов данных), в которой отражает 
    количество заёмщиков, долю и количество должников для каждого из сегментов 
    
    '''
        
    pivot_table = data.pivot_table(index=[column_name], values='debt', aggfunc=['count', 'sum', 'mean'])
    pivot_table.columns = ['all_credits', 'all_debtors', 'share_debtors']
    display(pivot_table)
    
display_pivot_table('children')

all_credits  all_debtors  share_debtors
children                                         
0               14091         1063       0.075438
1                4808          444       0.092346
2                2052          194       0.094542
3                 330           27       0.081818
4                  41            4       0.097561
5                   9            0       0.000000

**Вывод:** Из сводной таблицы видно, что лидеры по наличию задолженностей по возврату кредитов — заёмщики без детей:
- у заёмщиков без детей – 1 063;  
- среди родителей 1 ребёнка 444 должника; 
- 2 детей — 194 должника; 
- 3 детей – 27 должников; 
- 4 детей – 4 должника; 
- в категории родителей 5-ти детей должников нет.

Однако, для корректной оценки следует учитывать долю заёмщиков с задолженностями в каждой категории:
- у заёмщиков без детей – 0,075438 (~8%);  
- среди родителей 1 ребёнка – 0,092346 (~9%); 
- 2 детей — 0,094542 (~9%); 
- 3 детей – 0,081818 (~8%); 
- 4 детей – 0,097561 (~10%); 
- в категории родителей 5-ти детей должников нет, однако общее количество таких заёмщиков составляет всего 9 человек.

Как видно из данных отклонения среди категорией с самым большим количеством должников и самым маленьким - 2%. Из этого можно сделать вывод, что наличие детей не влияет на вероятность возврата кредита в срок. Значения долей заёмщиков с задолженностью по кредитам в каждой из категорий незначительно отличаются в относительном выражении к заёмщикам без задолженностей.  
  
При сравнении групп без детей с родителями можно отметить более существенное отклонение – у заёмщиков без детей самая низкая доля должников - 0,075438; а максимальное количество должников среди родителей 4 детей – 0,097561.   
При этом показатель группы родителей с 5-ю детьми можно не учитывать из-за маленького количества заёмщиков в выборке (9)


#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
# Применяем функцию вывода сводной таблицы к столбцу с id семейного положения
'''
соответствие семейного положения family_status_id:
    0 - женат/замужем
    1 - гражданский брак
    2 - вдовец / вдова
    3 - в разводе 
    4 - Не женат / не замужем

'''
display_pivot_table('family_status_id')

all_credits  all_debtors  share_debtors
family_status_id                                         
0                       12261          927       0.075606
1                        4134          385       0.093130
2                         951           63       0.066246
3                        1189           84       0.070648
4                        2796          273       0.097639

**Вывод:**
Лидеры по абсолютному количеству заёмщиков с задолженностями по возврату кредитов — заёмщики, состоящие в браке. Количество заёмщиков с задолженностю составило:
- у заёмщиков остоящих в браке – 927; 
- среди пар состоящих в гражданском браке 385 должников; 
- среди неженатых / незамужних — 273 должника; 
- самое меньшее количество должников среди разведенных и вдовых — 84 и 63 соответственно.

Однако, при расчёте относительных показателей к общему числу заёмщиков каждой категории:

- у заёмщиков состоящих в браке – 0,075606 (~8%); 
- среди пар состоящих в гражданском браке 0.093130 (~9%); 
- среди неженатых / незамужних — 0.097639 (~9%); 
- среди разведенных – 0.070648 (~7%)
- среди вдовых — 0.066246 (~7%).

Как видно из относительных данных разница в доле должников каждой из категорий не более 2%. Из этого можно сделать вывод, что семейное положение не влияет на вероятность возврата кредита в срок. Значения долей заёмщиков с задолженностью по кредитам в каждой из категорий незначительно отличаются в относительном выражении к заёмщикам без задолженностей.  

При этом отметим следующее: 
- у заёмщиков состоящих в браке – 0,075606 (~8%);
- среди пар состоящих в гражданском браке 0.093130 (~9%)
    
Отклонение между этими группами составляет 0,017524 (23% от состоящих в браке). Если сравнивать доли отклонений % между категориями, то мы можем присвоить более высокий рейтинг парам, состоящим в официальном браке по сравнении с состоящими в гражданском.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [25]:
# Применяем функцию вывода сводной таблицы к столбцу категорий уровня доходов
display_pivot_table('total_income_category')

all_credits  all_debtors  share_debtors
total_income_category                                         
A                               25            2       0.080000
B                             5014          354       0.070602
C                            15921         1353       0.084982
D                              349           21       0.060172
E                               22            2       0.090909

**Вывод:**  Заёмщики со средним доходом (50 001 - 200 000 руб.) ячейка 'C' — лидеры по количеству должников (1 353). 
Значительно меньше (более чем в три раза) просроченных долгов среди заёмщиков с уровнем дохода от 200 000 до 1 000 000 руб. — 354, почти в 17 раз меньше просроченных долгов у заёмщиков с доходами ниже среднего – 21.
В группах с самыми низкими и самыми высокими доходами по 2 заёмщика имели задолженности по кредитам.

Сравним доли должников в каждой из категорий:
- А: 0,080 (8%)
- В: 0,070602 (~7%)
- С: 0.084982 (~8%)
- D: 0.060172 (~6%)
- E: 0.090909 (~9%)

Как видно из относительных данных разница в доле должников каждой из категорий не более 3%. Из этого можно сделать вывод, что уровень дохода не влияет на вероятность возврата кредита в срок. Значения долей заёмщиков с задолженностью по кредитам в каждой из категорий незначительно отличаются в относительном выражении к заёмщикам без задолженностей.  

Следует отметить, что всегда стоит смотреть на количество записей в группах, чтобы не случилось ситуации, в которой будут делаться выводы на малом количестве данных (категория Е)

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [26]:
# Применяем функцию вывода сводной таблицы к столбцу категорий целей кредитования
display_pivot_table('purpose_category')

all_credits  all_debtors  share_debtors
purpose_category                                                 
операции с автомобилем           4279          400       0.093480
операции с недвижимостью        10751          780       0.072551
получение образования            3988          369       0.092528
проведение свадьбы               2313          183       0.079118

**Вывод:** Абсолютные значения количества заёмщиков с задолженностью по целям кредитов:

- операции с недвижимостью – 780;
- операции с автомобилем – 400 должников;
- на получение образования – 369 должников;
- на проведение свадьбы – 183.

Однако, при расчёте относительных показателей к общему числу заёмщиков каждой категории:

- операции с недвижимостью – 0.072551 (~7%);
- операции с автомобилем – 0.093480 (~9%);
- получение образования – 0.092528 (~9%);
- проведение свадьбы – 0.079118 (~7%).

Из относительных показателей видно, что и цели порлучения кредита не могут существенно влиять на вероятность своевременного возврата, разница в доле заёмщиков с задолженностью не более 2%.  

Но если сравнить  0.072551 и 0.093480, то разница в скоринге 
этих двух групп составляет уже \~28,8% (0.072551 - 0.093480 / 0.072551).  
    
С этой точки зрения группы заёмщиков с целями на операции с автомобилем и получение образования могут получать оценку существенно выше, чем на операции с недвижимостью и проведение свадьбы.

#### 3.5 Возможные причины появления пропусков в исходных данных.

*Ответ:*  В исходных данных датафрейма пропуски были обнаружены в столбцах с общим трудовым стажем 'days_employed' и ежемесячном доходе'total_income'. 

Такие пропуски могли возникнуть из-за недостаточного предоставления информации заёмщиками, неподтверждённой информации, отсутствии требований по их предоставлению для некоторых видов кредитов (например, товарные займы с минимальным пакетов подтверждающей информации).

Однако, если известны другие данные о потенциальном заёмщике, то мы можем работать с такими клиентами без ущерба для оценки рисков.

#### 3.6 Почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:*  При обнаружении пропусков наиболее разумно категоризировать заёмщиков по категориям связанным с такими пропусками, например уровень дохода напрямую зависит от типа занятости и заполнить пропуски медианными значениями каждой из категорий.
Медиана  —  это такое число в выборке, что ровно половина элементов больше него, а другая половина —  меньше. На медианное значение выборки не влияют строки с существенными отклонениями от типичных значений для каждой категории. Таким образом медианное значение характеризиует состояние выборки.

### Шаг 4: общий вывод.

Предоставлен таблица с данными о заёмщиках (21 525 строк), включающая в себя следующие столбцы:

- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита
    
     Пустые значения (2 174 строки) выявлены в столбцах со стажем ('days_employed') и размером дохода ('total_income'). 
    Для заполнения пропусков сгруппировала всех заёмщиков по типу занятости, рассчитала медианное значение в каждой из групп и указала их в пропущенных строках. Такое решение позволило не потерять важную информацию и не исказить общие данные для последующего анализа.  
    Далее были обработаны аномальные значения в данных столбца о количестве детей ('children'), выявлены строки с аномальными значениями (-1, 20). Строки с такими данными были удалены.
    Также были выявлены и удалены строки-дубликаты во избежание искажения итоговой оценки.
    Данные о заёмщиках объединила в категории по количеству детей, семейному положению, целям кредита и уровню дохода. Полученные категории оценила по столбцу о наличии задолженности по возврату кредитов ('debt').  
    
    
    В каждой из категорий рассчитаны:
    - общее количество заёмщиков;
    - общее количество заёмщиков, имеющих задолженность по возврату кредитов;
    - доля заёмщиков, имеющих задолженность по возврату кредитов.  
    
    
    Подробная информацию по результатам анализа и выводам представлена в соответствующих разделах.